File: src/config/ssot_config.py
#!/usr/bin/env python3
# AutoBot - AI-Powered Automation Platform
# Copyright (c) 2025 mrveiss
# Author: mrveiss
"""
SSOT Configuration Loader for Python
=====================================

Single Source of Truth - reads from .env file with AUTOBOT_ prefix.

This module provides typed configuration access across the entire backend.
All configuration values are defined in the master .env file and accessed
through Pydantic models for type safety and validation.

Usage:
    from src.config.ssot_config import get_config, config

    # Access VM IPs
    redis_ip = config.vm.redis
    backend_ip = config.vm.main

    # Access ports
    backend_port = config.port.backend

    # Access computed URLs
    backend_url = config.backend_url
    redis_url = config.redis_url

    # Access LLM settings
    default_model = config.llm.default_model

    # Access timeouts
    http_timeout = config.timeout.http

Issue: #601 - SSOT Phase 1: Foundation
Related: #599 - SSOT Configuration System Epic
"""

from __future__ import annotations

import os
from functools import lru_cache
from pathlib import Path
from typing import Optional

from pydantic import Field
from pydantic_settings import BaseSettings, SettingsConfigDict


# Determine project root for .env file location
def _find_project_root() -> Path:
    """Find the project root directory containing .env file."""
    current = Path(__file__).resolve()
    for parent in [current] + list(current.parents):
        if (parent / ".env").exists():
            return parent
    # Fallback to expected location
    return Path("/home/kali/Desktop/AutoBot")


PROJECT_ROOT = _find_project_root()

# Default model constants - single source of truth for fallback values
# These are used when .env doesn't specify a value
DEFAULT_LLM_MODEL = "mistral:7b-instruct"
DEFAULT_EMBEDDING_MODEL = "nomic-embed-text:latest"


class VMConfig(BaseSettings):
    """
    VM IP address configuration.

    Supports the 6-VM distributed architecture:
    - Main (WSL) - Backend API + VNC Desktop
    - Frontend (VM1) - Web interface
    - NPU Worker (VM2) - Hardware AI acceleration
    - Redis (VM3) - Data layer
    - AI Stack (VM4) - AI processing
    - Browser (VM5) - Web automation
    """

    model_config = SettingsConfigDict(
        env_file=str(PROJECT_ROOT / ".env"),
        env_file_encoding="utf-8",
        extra="ignore",
    )

    main: str = Field(default="172.16.168.20", alias="AUTOBOT_BACKEND_HOST")
    frontend: str = Field(default="172.16.168.21", alias="AUTOBOT_FRONTEND_HOST")
    npu: str = Field(default="172.16.168.22", alias="AUTOBOT_NPU_WORKER_HOST")
    redis: str = Field(default="172.16.168.23", alias="AUTOBOT_REDIS_HOST")
    aistack: str = Field(default="172.16.168.24", alias="AUTOBOT_AI_STACK_HOST")
    browser: str = Field(default="172.16.168.25", alias="AUTOBOT_BROWSER_SERVICE_HOST")
    ollama: str = Field(default="127.0.0.1", alias="AUTOBOT_OLLAMA_HOST")


class PortConfig(BaseSettings):
    """Service port configuration."""

    model_config = SettingsConfigDict(
        env_file=str(PROJECT_ROOT / ".env"),
        env_file_encoding="utf-8",
        extra="ignore",
    )

    backend: int = Field(default=8001, alias="AUTOBOT_BACKEND_PORT")
    frontend: int = Field(default=5173, alias="AUTOBOT_FRONTEND_PORT")
    redis: int = Field(default=6379, alias="AUTOBOT_REDIS_PORT")
    ollama: int = Field(default=11434, alias="AUTOBOT_OLLAMA_PORT")
    vnc: int = Field(default=6080, alias="AUTOBOT_VNC_PORT")
    browser: int = Field(default=3000, alias="AUTOBOT_BROWSER_SERVICE_PORT")
    aistack: int = Field(default=8080, alias="AUTOBOT_AI_STACK_PORT")
    npu: int = Field(default=8081, alias="AUTOBOT_NPU_WORKER_PORT")
    prometheus: int = Field(default=9090, alias="AUTOBOT_PROMETHEUS_PORT")
    grafana: int = Field(default=3000, alias="AUTOBOT_GRAFANA_PORT")


class LLMConfig(BaseSettings):
    """LLM model configuration."""

    model_config = SettingsConfigDict(
        env_file=str(PROJECT_ROOT / ".env"),
        env_file_encoding="utf-8",
        extra="ignore",
    )

    # Primary models - all use DEFAULT_LLM_MODEL constant for consistency
    default_model: str = Field(
        default=DEFAULT_LLM_MODEL, alias="AUTOBOT_DEFAULT_LLM_MODEL"
    )
    embedding_model: str = Field(
        default=DEFAULT_EMBEDDING_MODEL, alias="AUTOBOT_EMBEDDING_MODEL"
    )
    classification_model: str = Field(
        default=DEFAULT_LLM_MODEL, alias="AUTOBOT_CLASSIFICATION_MODEL"
    )
    reasoning_model: str = Field(
        default=DEFAULT_LLM_MODEL, alias="AUTOBOT_REASONING_MODEL"
    )
    rag_model: str = Field(default=DEFAULT_LLM_MODEL, alias="AUTOBOT_RAG_MODEL")
    coding_model: str = Field(
        default=DEFAULT_LLM_MODEL, alias="AUTOBOT_CODING_MODEL"
    )

    # Agent/workflow models - all use DEFAULT_LLM_MODEL constant
    orchestrator_model: str = Field(
        default=DEFAULT_LLM_MODEL, alias="AUTOBOT_ORCHESTRATOR_MODEL"
    )
    agent_model: str = Field(
        default=DEFAULT_LLM_MODEL, alias="AUTOBOT_AGENT_MODEL"
    )
    research_model: str = Field(
        default=DEFAULT_LLM_MODEL, alias="AUTOBOT_RESEARCH_MODEL"
    )
    analysis_model: str = Field(
        default=DEFAULT_LLM_MODEL, alias="AUTOBOT_ANALYSIS_MODEL"
    )
    planning_model: str = Field(
        default=DEFAULT_LLM_MODEL, alias="AUTOBOT_PLANNING_MODEL"
    )

    # Timeout (seconds)
    timeout: int = Field(default=30, alias="AUTOBOT_LLM_TIMEOUT")

    # Provider (for future multi-provider support)
    provider: str = Field(default="ollama", alias="AUTOBOT_LLM_PROVIDER")


class TimeoutConfig(BaseSettings):
    """
    Timeout configuration (in milliseconds for HTTP, seconds for LLM).

    Note: Frontend uses milliseconds, backend uses seconds for LLM.
    The config stores values as they appear in .env for consistency.
    """

    model_config = SettingsConfigDict(
        env_file=str(PROJECT_ROOT / ".env"),
        env_file_encoding="utf-8",
        extra="ignore",
    )

    # API timeouts (milliseconds)
    api: int = Field(default=10000, alias="AUTOBOT_API_TIMEOUT")
    api_retry_attempts: int = Field(default=3, alias="AUTOBOT_API_RETRY_ATTEMPTS")
    api_retry_delay: int = Field(default=1000, alias="AUTOBOT_API_RETRY_DELAY")

    # LLM timeout (seconds)
    llm: int = Field(default=30, alias="AUTOBOT_LLM_TIMEOUT")

    # Health check timeout (seconds)
    health_check: int = Field(default=3, alias="AUTOBOT_HEALTH_CHECK_TIMEOUT")

    # WebSocket timeout (seconds)
    websocket: int = Field(default=30, alias="AUTOBOT_WEBSOCKET_TIMEOUT")

    @property
    def api_seconds(self) -> float:
        """Get API timeout in seconds."""
        return self.api / 1000.0

    @property
    def http(self) -> int:
        """Alias for api timeout (milliseconds)."""
        return self.api

    @property
    def http_seconds(self) -> float:
        """Get HTTP timeout in seconds."""
        return self.api / 1000.0


class RedisConfig(BaseSettings):
    """Redis database configuration."""

    model_config = SettingsConfigDict(
        env_file=str(PROJECT_ROOT / ".env"),
        env_file_encoding="utf-8",
        extra="ignore",
    )

    # Enable/disable Redis
    enabled: bool = Field(default=True, alias="AUTOBOT_REDIS_ENABLED")

    # Database assignments
    db_main: int = Field(default=0, alias="AUTOBOT_REDIS_DB_MAIN")
    db_knowledge: int = Field(default=1, alias="AUTOBOT_REDIS_DB_KNOWLEDGE")
    db_prompts: int = Field(default=2, alias="AUTOBOT_REDIS_DB_PROMPTS")
    db_agents: int = Field(default=3, alias="AUTOBOT_REDIS_DB_AGENTS")
    db_metrics: int = Field(default=4, alias="AUTOBOT_REDIS_DB_METRICS")
    db_cache: int = Field(default=5, alias="AUTOBOT_REDIS_DB_CACHE")
    db_sessions: int = Field(default=6, alias="AUTOBOT_REDIS_DB_SESSIONS")
    db_tasks: int = Field(default=7, alias="AUTOBOT_REDIS_DB_TASKS")
    db_logs: int = Field(default=8, alias="AUTOBOT_REDIS_DB_LOGS")
    db_temp: int = Field(default=9, alias="AUTOBOT_REDIS_DB_TEMP")
    db_backup: int = Field(default=10, alias="AUTOBOT_REDIS_DB_BACKUP")
    db_testing: int = Field(default=15, alias="AUTOBOT_REDIS_DB_TESTING")

    # Security
    password: Optional[str] = Field(default=None, alias="AUTOBOT_REDIS_PASSWORD")


class FeatureConfig(BaseSettings):
    """Feature flags configuration."""

    model_config = SettingsConfigDict(
        env_file=str(PROJECT_ROOT / ".env"),
        env_file_encoding="utf-8",
        extra="ignore",
    )

    unified_config: bool = Field(default=True, alias="AUTOBOT_USE_UNIFIED_CONFIG")
    semantic_chunking: bool = Field(default=True, alias="AUTOBOT_SEMANTIC_CHUNKING")
    debug_mode: bool = Field(default=False, alias="AUTOBOT_DEBUG_MODE")
    hot_reload: bool = Field(default=True, alias="AUTOBOT_HOT_RELOAD")
    single_user_mode: bool = Field(default=True, alias="AUTOBOT_SINGLE_USER_MODE")


class AutoBotConfig(BaseSettings):
    """
    Master configuration - SINGLE SOURCE OF TRUTH.

    This class aggregates all configuration sections and provides
    computed properties for commonly used URLs.

    Usage:
        from src.config.ssot_config import get_config

        config = get_config()
        backend = config.backend_url  # http://172.16.168.20:8001
        redis = config.redis_url  # redis://172.16.168.23:6379
        model = config.llm.default_model  # mistral:7b-instruct
    """

    model_config = SettingsConfigDict(
        env_file=str(PROJECT_ROOT / ".env"),
        env_file_encoding="utf-8",
        extra="ignore",
    )

    # Sub-configurations
    vm: VMConfig = Field(default_factory=VMConfig)
    port: PortConfig = Field(default_factory=PortConfig)
    llm: LLMConfig = Field(default_factory=LLMConfig)
    timeout: TimeoutConfig = Field(default_factory=TimeoutConfig)
    redis: RedisConfig = Field(default_factory=RedisConfig)
    feature: FeatureConfig = Field(default_factory=FeatureConfig)

    # Top-level settings
    deployment_mode: str = Field(default="distributed", alias="AUTOBOT_DEPLOYMENT_MODE")
    environment: str = Field(default="development", alias="AUTOBOT_ENVIRONMENT")
    debug: bool = Field(default=False, alias="AUTOBOT_DEBUG")
    log_level: str = Field(default="INFO", alias="AUTOBOT_LOG_LEVEL")

    # Computed URL properties for backward compatibility
    @property
    def backend_url(self) -> str:
        """Get the full backend API URL."""
        return f"http://{self.vm.main}:{self.port.backend}"

    @property
    def frontend_url(self) -> str:
        """Get the full frontend URL."""
        return f"http://{self.vm.frontend}:{self.port.frontend}"

    @property
    def redis_url(self) -> str:
        """Get the full Redis URL (without password)."""
        return f"redis://{self.vm.redis}:{self.port.redis}"

    @property
    def redis_url_with_auth(self) -> str:
        """Get the full Redis URL with password if configured."""
        if self.redis.password:
            return f"redis://:{self.redis.password}@{self.vm.redis}:{self.port.redis}"
        return self.redis_url

    @property
    def ollama_url(self) -> str:
        """Get the full Ollama API URL."""
        return f"http://{self.vm.ollama}:{self.port.ollama}"

    @property
    def websocket_url(self) -> str:
        """Get the WebSocket URL for real-time communication."""
        return f"ws://{self.vm.main}:{self.port.backend}/ws"

    @property
    def aistack_url(self) -> str:
        """Get the AI Stack service URL."""
        return f"http://{self.vm.aistack}:{self.port.aistack}"

    @property
    def npu_worker_url(self) -> str:
        """Get the NPU Worker service URL."""
        return f"http://{self.vm.npu}:{self.port.npu}"

    @property
    def browser_service_url(self) -> str:
        """Get the Browser automation service URL."""
        return f"http://{self.vm.browser}:{self.port.browser}"

    @property
    def vnc_url(self) -> str:
        """Get the VNC desktop URL."""
        return f"http://{self.vm.main}:{self.port.vnc}/vnc.html"

    def get_redis_url_for_db(self, db_number: int) -> str:
        """Get Redis URL for a specific database number."""
        base = self.redis_url_with_auth
        return f"{base}/{db_number}"

    def get_service_url(self, service_name: str) -> Optional[str]:
        """
        Get service URL by name.

        Args:
            service_name: Name of service (backend, frontend, redis, ollama, etc.)

        Returns:
            Service URL or None if not found
        """
        url_map = {
            "backend": self.backend_url,
            "frontend": self.frontend_url,
            "redis": self.redis_url,
            "ollama": self.ollama_url,
            "websocket": self.websocket_url,
            "aistack": self.aistack_url,
            "ai_stack": self.aistack_url,
            "npu": self.npu_worker_url,
            "npu_worker": self.npu_worker_url,
            "browser": self.browser_service_url,
            "vnc": self.vnc_url,
        }
        return url_map.get(service_name.lower())

    def get_vm_ip(self, vm_name: str) -> Optional[str]:
        """
        Get VM IP address by name.

        Args:
            vm_name: Name of VM (main, frontend, redis, etc.)

        Returns:
            IP address or None if not found
        """
        vm_map = {
            "main": self.vm.main,
            "frontend": self.vm.frontend,
            "npu": self.vm.npu,
            "npu_worker": self.vm.npu,
            "redis": self.vm.redis,
            "aistack": self.vm.aistack,
            "ai_stack": self.vm.aistack,
            "browser": self.vm.browser,
            "ollama": self.vm.ollama,
        }
        return vm_map.get(vm_name.lower())


@lru_cache(maxsize=1)
def get_config() -> AutoBotConfig:
    """
    Get singleton configuration instance.

    Uses LRU cache to ensure only one instance is created.
    The configuration is loaded from .env file on first access.

    Returns:
        AutoBotConfig instance with all configuration loaded
    """
    return AutoBotConfig()


def reload_config() -> AutoBotConfig:
    """
    Force reload of configuration.

    Clears the cached config and creates a new instance.
    Use this when .env file has changed and you need fresh values.

    Returns:
        Fresh AutoBotConfig instance
    """
    get_config.cache_clear()
    return get_config()


# Convenience export - lazy singleton
# This allows: from src.config.ssot_config import config
# while still supporting get_config() for explicit access
class _ConfigProxy:
    """Lazy proxy for config singleton."""

    def __getattr__(self, name):
        return getattr(get_config(), name)


config = _ConfigProxy()


# Backward compatibility exports
# These provide drop-in compatibility with NetworkConstants usage patterns
def get_backend_url() -> str:
    """Get backend URL (backward compatibility)."""
    return get_config().backend_url


def get_redis_url() -> str:
    """Get Redis URL (backward compatibility)."""
    return get_config().redis_url


def get_ollama_url() -> str:
    """Get Ollama URL (backward compatibility)."""
    return get_config().ollama_url


def get_default_llm_model() -> str:
    """Get default LLM model name (backward compatibility)."""
    return get_config().llm.default_model


# Export commonly used values at module level for convenience
__all__ = [
    "AutoBotConfig",
    "VMConfig",
    "PortConfig",
    "LLMConfig",
    "TimeoutConfig",
    "RedisConfig",
    "FeatureConfig",
    "get_config",
    "reload_config",
    "config",
    # Backward compatibility
    "get_backend_url",
    "get_redis_url",
    "get_ollama_url",
    "get_default_llm_model",
    "PROJECT_ROOT",
]

File: autobot-vue/src/config/ssot-config.ts
/**
 * SSOT Configuration Loader for TypeScript
 * =========================================
 *
 * Single Source of Truth - reads from Vite environment variables.
 *
 * This module provides typed configuration access across the entire frontend.
 * All configuration values are defined in the master .env file and mapped
 * to VITE_* prefixed variables for Vite to expose them.
 *
 * Usage:
 *   import { getConfig, config } from '@/config/ssot-config';
 *
 *   // Access VM IPs
 *   const redisIp = config.vm.redis;
 *   const backendIp = config.vm.main;
 *
 *   // Access ports
 *   const backendPort = config.port.backend;
 *
 *   // Access computed URLs
 *   const backendUrl = config.backendUrl;
 *   const websocketUrl = config.websocketUrl;
 *
 *   // Access timeouts
 *   const apiTimeout = config.timeout.api;
 *
 * Issue: #601 - SSOT Phase 1: Foundation
 * Related: #599 - SSOT Configuration System Epic
 *
 * AutoBot - AI-Powered Automation Platform
 * Copyright (c) 2025 mrveiss
 * Author: mrveiss
 */

// =============================================================================
// Type Definitions
// =============================================================================

/**
 * VM IP address configuration.
 * Supports the 6-VM distributed architecture.
 */
export interface VMConfig {
  /** Main machine (WSL) - Backend API + VNC Desktop */
  main: string;
  /** VM1 Frontend - Web interface */
  frontend: string;
  /** VM2 NPU Worker - Hardware AI acceleration */
  npu: string;
  /** VM3 Redis - Data layer */
  redis: string;
  /** VM4 AI Stack - AI processing */
  aistack: string;
  /** VM5 Browser - Web automation */
  browser: string;
  /** Ollama host (typically localhost) */
  ollama: string;
  /** SLM Backend (Service Lifecycle Manager) - Issue #725 */
  slm: string;
}

/**
 * Service port configuration.
 */
export interface PortConfig {
  backend: number;
  frontend: number;
  redis: number;
  ollama: number;
  vnc: number;
  browser: number;
  aistack: number;
  npu: number;
  prometheus: number;
  grafana: number;
  /** SLM Backend port - Issue #725 */
  slm: number;
}

/**
 * LLM model configuration.
 */
export interface LLMConfig {
  defaultModel: string;
  embeddingModel: string;
  provider: string;
  timeout: number;
}

/**
 * Timeout configuration (in milliseconds).
 */
export interface TimeoutConfig {
  /** API request timeout (ms) */
  api: number;
  /** Knowledge/LLM operations timeout (ms) */
  knowledge: number;
  /** API retry attempts */
  retryAttempts: number;
  /** Delay between retries (ms) */
  retryDelay: number;
  /** WebSocket timeout (ms) */
  websocket: number;
}

/**
 * VNC configuration for desktop streaming.
 */
export interface VNCConfig {
  desktop: {
    host: string;
    port: number;
    password: string;
  };
  terminal: {
    host: string;
    port: number;
    password: string;
  };
  playwright: {
    host: string;
    port: number;
    password: string;
  };
}

/**
 * Feature flags configuration.
 */
export interface FeatureConfig {
  debug: boolean;
  rum: boolean;
  cacheDisabled: boolean;
}

// =============================================================================
// Permission System v2 Types - Claude Code Style
// Issue: Permission System v2 - Claude Code Style
// =============================================================================

/**
 * Permission modes matching backend PermissionMode enum.
 */
export type PermissionMode =
  | 'default'
  | 'acceptEdits'
  | 'plan'
  | 'dontAsk'
  | 'bypassPermissions';

/**
 * Permission actions matching backend PermissionAction enum.
 */
export type PermissionAction = 'allow' | 'ask' | 'deny';

/**
 * Permission rule interface.
 */
export interface PermissionRule {
  tool: string;
  pattern: string;
  action: PermissionAction;
  description: string;
}

/**
 * Approval record interface for project memory.
 */
export interface ApprovalRecord {
  pattern: string;
  tool: string;
  risk_level: string;
  user_id: string;
  created_at: number;
  original_command: string;
  comment?: string;
}

/**
 * Permission system status interface.
 */
export interface PermissionStatus {
  enabled: boolean;
  mode: PermissionMode;
  approval_memory_enabled: boolean;
  approval_memory_ttl_days: number;
  rules_file: string;
  rules_count: {
    allow: number;
    ask: number;
    deny: number;
    total: number;
  };
}

/**
 * Permission configuration interface.
 */
export interface PermissionConfig {
  /** Whether permission system v2 is enabled */
  enabled: boolean;
  /** Current permission mode */
  mode: PermissionMode;
  /** Whether approval memory is enabled */
  approvalMemoryEnabled: boolean;
  /** TTL for approval memory in seconds */
  approvalMemoryTtl: number;
  /** Path to rules YAML file */
  rulesFile: string;
  /** Admin-only modes that require elevated privileges */
  adminOnlyModes: PermissionMode[];
}

/**
 * Match result from permission checker.
 */
export type MatchResult = 'allow' | 'ask' | 'deny' | 'default';

/**
 * Complete AutoBot configuration.
 */
export interface AutoBotConfig {
  // Sub-configurations
  vm: VMConfig;
  port: PortConfig;
  llm: LLMConfig;
  timeout: TimeoutConfig;
  vnc: VNCConfig;
  feature: FeatureConfig;

  // Top-level settings
  deploymentMode: string;
  environment: string;
  debug: boolean;
  httpProtocol: string;

  // Computed URLs (readonly getters)
  readonly backendUrl: string;
  readonly frontendUrl: string;
  readonly redisUrl: string;
  readonly ollamaUrl: string;
  readonly websocketUrl: string;
  readonly aistackUrl: string;
  readonly npuWorkerUrl: string;
  readonly browserServiceUrl: string;
  readonly vncUrl: string;
  /** SLM Backend URL - Issue #725 */
  readonly slmUrl: string;
}

// =============================================================================
// Environment Variable Helpers
// =============================================================================

/**
 * Get string environment variable with fallback.
 */
function getEnv(key: string, defaultValue: string): string {
  const value = import.meta.env[key];
  if (value === undefined || value === null || value === '') {
    return defaultValue;
  }
  return String(value);
}

/**
 * Get numeric environment variable with fallback.
 */
function getEnvNumber(key: string, defaultValue: number): number {
  const value = import.meta.env[key];
  if (value === undefined || value === null || value === '') {
    return defaultValue;
  }
  const parsed = parseInt(String(value), 10);
  return isNaN(parsed) ? defaultValue : parsed;
}

/**
 * Get boolean environment variable with fallback.
 */
function getEnvBoolean(key: string, defaultValue: boolean): boolean {
  const value = import.meta.env[key];
  if (value === undefined || value === null || value === '') {
    return defaultValue;
  }
  const strValue = String(value).toLowerCase();
  return strValue === 'true' || strValue === '1' || strValue === 'yes';
}

// =============================================================================
// Configuration Builder
// =============================================================================

/**
 * Build the complete configuration from environment variables.
 * This is called once on first access.
 */
function buildConfig(): AutoBotConfig {
  // VM IP addresses
  const vm: VMConfig = {
    main: getEnv('VITE_BACKEND_HOST', '172.16.168.20'),
    frontend: getEnv('VITE_FRONTEND_HOST', '172.16.168.21'),
    npu: getEnv('VITE_NPU_WORKER_HOST', '172.16.168.22'),
    redis: getEnv('VITE_REDIS_HOST', '172.16.168.23'),
    aistack: getEnv('VITE_AI_STACK_HOST', '172.16.168.24'),
    browser: getEnv('VITE_BROWSER_HOST', '172.16.168.25'),
    ollama: getEnv('VITE_OLLAMA_HOST', '127.0.0.1'),
    slm: getEnv('VITE_SLM_HOST', '172.16.168.19'),
  };

  // Service ports
  const port: PortConfig = {
    backend: getEnvNumber('VITE_BACKEND_PORT', 8001),
    frontend: getEnvNumber('VITE_FRONTEND_PORT', 5173),
    redis: getEnvNumber('VITE_REDIS_PORT', 6379),
    ollama: getEnvNumber('VITE_OLLAMA_PORT', 11434),
    vnc: getEnvNumber('VITE_DESKTOP_VNC_PORT', 6080),
    browser: getEnvNumber('VITE_BROWSER_PORT', 3000),
    aistack: getEnvNumber('VITE_AI_STACK_PORT', 8080),
    npu: getEnvNumber('VITE_NPU_WORKER_PORT', 8081),
    prometheus: getEnvNumber('VITE_PROMETHEUS_PORT', 9090),
    grafana: getEnvNumber('VITE_GRAFANA_PORT', 3000),
    slm: getEnvNumber('VITE_SLM_PORT', 8000),
  };

  // LLM configuration
  const llm: LLMConfig = {
    defaultModel: getEnv('VITE_LLM_DEFAULT_MODEL', 'mistral:7b-instruct'),
    embeddingModel: getEnv('VITE_LLM_EMBEDDING_MODEL', 'nomic-embed-text:latest'),
    provider: getEnv('VITE_LLM_PROVIDER', 'ollama'),
    timeout: getEnvNumber('VITE_LLM_TIMEOUT', 30000),
  };

  // Timeout configuration (all in milliseconds)
  const timeout: TimeoutConfig = {
    api: getEnvNumber('VITE_API_TIMEOUT', 60000),
    knowledge: getEnvNumber('VITE_KNOWLEDGE_TIMEOUT', 300000),
    retryAttempts: getEnvNumber('VITE_API_RETRY_ATTEMPTS', 3),
    retryDelay: getEnvNumber('VITE_API_RETRY_DELAY', 1000),
    websocket: getEnvNumber('VITE_WEBSOCKET_TIMEOUT', 30000),
  };

  // VNC configuration
  const vnc: VNCConfig = {
    desktop: {
      host: getEnv('VITE_DESKTOP_VNC_HOST', vm.main),
      port: getEnvNumber('VITE_DESKTOP_VNC_PORT', 6080),
      password: getEnv('VITE_DESKTOP_VNC_PASSWORD', 'autobot'),
    },
    terminal: {
      host: getEnv('VITE_TERMINAL_VNC_HOST', vm.main),
      port: getEnvNumber('VITE_TERMINAL_VNC_PORT', 6080),
      password: getEnv('VITE_TERMINAL_VNC_PASSWORD', 'autobot'),
    },
    playwright: {
      host: getEnv('VITE_PLAYWRIGHT_VNC_HOST', vm.browser),
      port: getEnvNumber('VITE_PLAYWRIGHT_VNC_PORT', 6081),
      password: getEnv('VITE_PLAYWRIGHT_VNC_PASSWORD', 'playwright'),
    },
  };

  // Feature flags
  const feature: FeatureConfig = {
    debug: getEnvBoolean('VITE_ENABLE_DEBUG', false),
    rum: getEnvBoolean('VITE_ENABLE_RUM', true),
    cacheDisabled: getEnvBoolean('VITE_DISABLE_CACHE', false),
  };

  // Top-level settings
  const deploymentMode = getEnv('VITE_DEPLOYMENT_MODE', 'distributed');
  const environment = getEnv('VITE_ENVIRONMENT', 'development');
  const debug = getEnvBoolean('VITE_DEBUG', false);
  const httpProtocol = getEnv('VITE_HTTP_PROTOCOL', 'http');

  // Build the config object with computed URL getters
  const configObj: AutoBotConfig = {
    vm,
    port,
    llm,
    timeout,
    vnc,
    feature,
    deploymentMode,
    environment,
    debug,
    httpProtocol,

    // Computed URLs as getters
    get backendUrl(): string {
      return `${httpProtocol}://${vm.main}:${port.backend}`;
    },

    get frontendUrl(): string {
      return `${httpProtocol}://${vm.frontend}:${port.frontend}`;
    },

    get redisUrl(): string {
      return `redis://${vm.redis}:${port.redis}`;
    },

    get ollamaUrl(): string {
      return `${httpProtocol}://${vm.ollama}:${port.ollama}`;
    },

    get websocketUrl(): string {
      const wsProtocol = httpProtocol === 'https' ? 'wss' : 'ws';
      return `${wsProtocol}://${vm.main}:${port.backend}/ws`;
    },

    get aistackUrl(): string {
      return `${httpProtocol}://${vm.aistack}:${port.aistack}`;
    },

    get npuWorkerUrl(): string {
      return `${httpProtocol}://${vm.npu}:${port.npu}`;
    },

    get browserServiceUrl(): string {
      return `${httpProtocol}://${vm.browser}:${port.browser}`;
    },

    get vncUrl(): string {
      return `${httpProtocol}://${vm.main}:${port.vnc}/vnc.html`;
    },

    get slmUrl(): string {
      return `${httpProtocol}://${vm.slm}:${port.slm}`;
    },
  };

  return configObj;
}

// =============================================================================
// Singleton Configuration
// =============================================================================

/** Cached configuration instance */
let _config: AutoBotConfig | null = null;

/**
 * Get the singleton configuration instance.
 *
 * Configuration is loaded from Vite environment variables on first access.
 * Subsequent calls return the cached instance.
 *
 * @returns AutoBotConfig instance
 */
export function getConfig(): AutoBotConfig {
  if (_config === null) {
    _config = buildConfig();
  }
  return _config;
}

/**
 * Force reload of configuration.
 *
 * Clears the cached config and creates a new instance.
 * Use this when environment has changed dynamically.
 *
 * @returns Fresh AutoBotConfig instance
 */
export function reloadConfig(): AutoBotConfig {
  _config = null;
  return getConfig();
}

/**
 * Get a service URL by name.
 *
 * @param serviceName - Name of service (backend, frontend, redis, etc.)
 * @returns Service URL or undefined if not found
 */
export function getServiceUrl(serviceName: string): string | undefined {
  const cfg = getConfig();
  const urlMap: Record<string, string> = {
    backend: cfg.backendUrl,
    frontend: cfg.frontendUrl,
    redis: cfg.redisUrl,
    ollama: cfg.ollamaUrl,
    websocket: cfg.websocketUrl,
    aistack: cfg.aistackUrl,
    ai_stack: cfg.aistackUrl,
    npu: cfg.npuWorkerUrl,
    npu_worker: cfg.npuWorkerUrl,
    browser: cfg.browserServiceUrl,
    vnc: cfg.vncUrl,
    slm: cfg.slmUrl,
  };
  return urlMap[serviceName.toLowerCase()];
}

/**
 * Get a VM IP address by name.
 *
 * @param vmName - Name of VM (main, frontend, redis, etc.)
 * @returns IP address or undefined if not found
 */
export function getVmIp(vmName: string): string | undefined {
  const cfg = getConfig();
  const vmMap: Record<string, string> = {
    main: cfg.vm.main,
    frontend: cfg.vm.frontend,
    npu: cfg.vm.npu,
    npu_worker: cfg.vm.npu,
    redis: cfg.vm.redis,
    aistack: cfg.vm.aistack,
    ai_stack: cfg.vm.aistack,
    browser: cfg.vm.browser,
    ollama: cfg.vm.ollama,
    slm: cfg.vm.slm,
  };
  return vmMap[vmName.toLowerCase()];
}

// =============================================================================
// Default Export
// =============================================================================

/**
 * Lazy-loaded configuration singleton.
 *
 * Usage:
 *   import config from '@/config/ssot-config';
 *   console.log(config.backendUrl);
 */
const config = getConfig();
export default config;

// =============================================================================
// Backward Compatibility Exports
// =============================================================================

/**
 * Get backend URL (backward compatibility).
 */
export function getBackendUrl(): string {
  return getConfig().backendUrl;
}

/**
 * Get WebSocket URL (backward compatibility).
 */
export function getWebsocketUrl(): string {
  return getConfig().websocketUrl;
}

/**
 * Get Backend WebSocket URL base (without /ws suffix).
 * Used by SSHTerminal and other WebSocket-based components that need
 * to build their own WebSocket paths (e.g., /api/terminal/ws/ssh/{host_id}).
 *
 * Note: This differs from getWebsocketUrl() which includes /ws suffix
 * for the main chat WebSocket endpoint.
 */
export function getBackendWsUrl(): string {
  const backendUrl = getConfig().backendUrl;
  // Convert http(s) to ws(s) protocol
  return backendUrl.replace(/^http/, 'ws');
}

/**
 * Get API timeout in milliseconds (backward compatibility).
 */
export function getApiTimeout(): number {
  return getConfig().timeout.api;
}

/**
 * Get SLM Backend URL (Issue #725).
 */
export function getSLMUrl(): string {
  return getConfig().slmUrl;
}
